File: long_questions/DSA/Leetcode/golang/25_String_Parsing_Stacks/020_Valid_Parentheses.go
package main

import "fmt"

// 20. Valid Parentheses
// Time: O(N), Space: O(N)
func isValid(s string) bool {
	stack := []rune{}
	
	// Map closing brackets to opening brackets
	bracketMap := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}
	
	for _, char := range s {
		// If it's an opening bracket, push to stack
		switch char {
		case '(', '{', '[':
			stack = append(stack, char)
		case ')', '}', ']':
			// If stack is empty or top doesn't match, return false
			if len(stack) == 0 || stack[len(stack)-1] != bracketMap[char] {
				return false
			}
			stack = stack[:len(stack)-1]
		default:
			// Invalid character
			return false
		}
	}
	
	// Stack should be empty for valid string
	return len(stack) == 0
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Stack-Based Bracket Matching
- **Stack Usage**: Push opening brackets, pop on matching closing
- **Bracket Mapping**: Map closing brackets to opening brackets
- **LIFO Principle**: Last opened bracket must be closed first
- **Validation**: String valid if stack empty at end

## 2. PROBLEM CHARACTERISTICS
- **Nested Structure**: Brackets can be nested within each other
- **Matching Pairs**: Each opening has specific closing bracket
- **Order Validation**: Brackets must close in correct order
- **Complete Matching**: All brackets must be properly paired

## 3. SIMILAR PROBLEMS
- Generate Parentheses (LeetCode 22) - Generate valid combinations
- Longest Valid Parentheses (LeetCode 32) - Find longest valid substring
- Remove Invalid Parentheses (LeetCode 301) - Remove minimum to make valid
- Score of Parentheses (LeetCode 856) - Calculate score based on nesting

## 4. KEY OBSERVATIONS
- **Stack Natural Fit**: LIFO property matches bracket nesting
- **Early Validation**: Can detect invalid immediately on mismatch
- **Empty Stack Check**: Final validation requires empty stack
- **Character Validation**: Need to handle invalid characters

## 5. VARIATIONS & EXTENSIONS
- **Different Bracket Types**: Additional bracket types (angle brackets)
- **Multiple Strings**: Validate multiple strings simultaneously
- **Error Reporting**: Return specific error types/positions
- **Partial Validation**: Find longest valid substring

## 6. INTERVIEW INSIGHTS
- Always clarify: "Bracket types? Invalid characters? Empty string?"
- Edge cases: empty string, single bracket, invalid characters
- Time complexity: O(N) where N=string length
- Space complexity: O(N) in worst case (all opening brackets)
- Key insight: stack perfectly models nested structure

## 7. COMMON MISTAKES
- Not handling empty stack before popping
- Wrong bracket mapping (closing to opening)
- Not checking final stack emptiness
- Missing invalid character handling
- Off-by-one errors in stack indexing

## 8. OPTIMIZATION STRATEGIES
- **Basic Stack**: O(N) time, O(N) space - optimal
- **Early Termination**: Stop on first invalid character
- **Character Filtering**: Pre-filter valid characters
- **Memory Pool**: Reuse stack for multiple validations

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like matching nested containers:**
- You have containers that can contain other containers
- Each container type has specific lid (bracket pair)
- Must close containers in reverse order of opening
- Like Russian nesting dolls - smallest opened last
- Stack naturally models this nesting behavior

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: String containing brackets
2. **Goal**: Determine if brackets are properly matched
3. **Rules**: Each opening must have matching closing in correct order
4. **Output**: Boolean indicating validity

#### Phase 2: Key Insight Recognition
- **"Stack natural fit"** → LIFO matches bracket nesting
- **"Bracket mapping"** → Need to map closing to opening
- **"Early validation"** → Can detect errors immediately
- **"Final check"** → Stack must be empty at end

#### Phase 3: Strategy Development
```
Human thought process:
"I need to validate bracket matching.
This is perfect for stack usage:

Stack Approach:
1. Iterate through each character
2. If opening bracket: push to stack
3. If closing bracket:
   - Check if stack empty (invalid)
   - Pop top and check if matches
   - If no match: invalid
4. After processing all characters:
   - Stack must be empty for valid string

This naturally handles nesting and order!"
```

#### Phase 4: Edge Case Handling
- **Empty string**: Return true (no brackets to validate)
- **Single bracket**: Return false (cannot be matched)
- **Invalid characters**: Return false or clarify handling
- **All opening**: Return false (stack not empty at end)

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
String: "({[]})"

Human thinking:
"Step 1: '(' → opening, push to stack
Stack: [ ( ]

Step 2: '{' → opening, push to stack  
Stack: [ (, { ]

Step 3: '[' → opening, push to stack
Stack: [ (, {, [ ]

Step 4: ']' → closing, check top
Top is '[', matches ']' ✓ Pop
Stack: [ (, { ]

Step 5: '}' → closing, check top
Top is '{', matches '}' ✓ Pop
Stack: [ ( ]

Step 6: ')' → closing, check top
Top is '(', matches ')' ✓ Pop
Stack: [ ] (empty)

Step 7: End of string, stack empty ✓ Valid!"
```

#### Phase 6: Intuition Validation
- **Why stack works**: LIFO matches nested bracket behavior
- **Why mapping needed**: Need to know which opening matches closing
- **Why early check works**: Can detect invalid immediately
- **Why final check needed**: All openings must be closed

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just count?"** → Counting doesn't capture order
2. **"Should I use recursion?"** → Stack is more efficient
3. **"What about other characters?"** → Clarify handling policy
4. **"Can I optimize further?"** → O(N) is already optimal
5. **"What about different bracket types?"** → Extend mapping approach

### Real-World Analogy
**Like checking if a document has properly nested tags:**
- HTML/XML tags must be properly nested and closed
- Each opening tag has specific closing tag
- Tags can be nested within other tags
- Like validating HTML structure
- Stack naturally tracks which tag we're currently inside

### Human-Readable Pseudocode
```
function isValidParentheses(s):
    stack = empty stack
    bracketMap = {
        ')': '(',
        '}': '{',
        ']': '['
    }
    
    for char in s:
        if char is opening bracket:
            stack.push(char)
        else if char is closing bracket:
            if stack is empty:
                return false
            top = stack.pop()
            if top != bracketMap[char]:
                return false
        else:
            return false // invalid character
    
    return stack.isEmpty()
```

### Execution Visualization

### Example: String = "({[]})"
```
Character Processing:
Index: 0  1  2  3  4  5  6
Char:  (  {  [  ]  }  )
Stack: [ (] [({] [({[ [({] [({] [({] [({] [] 

Step-by-step:
1. '(' → opening → push: [ ( ]
2. '{' → opening → push: [ (, { ]
3. '[' → opening → push: [ (, {, [ ]
4. ']' → closing → top is '[', matches ✓ pop: [ (, { ]
5. '}' → closing → top is '{', matches ✓ pop: [ ( ]
6. ')' → closing → top is '(', matches ✓ pop: [ ]
7. End → stack empty ✓ Valid!

Final Result: true ✓
```

### Key Visualization Points:
- **Stack Growth**: Push all opening brackets
- **Stack Shrink**: Pop on successful matches
- **Order Validation**: LIFO ensures correct closing order
- **Final Check**: Empty stack means all matched

### Memory Layout Visualization:
```
Stack State Evolution:
Initial:  []
After '(':  [ ( ]
After '{':  [ (, { ]
After '[':  [ (, {, [ ]
After ']':  [ (, { ]    (popped '[')
After '}':  [ ( ]        (popped '{')
After ')':  [ ]            (popped '(')
Final:    []            (empty ✓)

Bracket Mapping:
')' → '('
'}' → '{'
']' → '['

Validation Steps:
1. Check character type
2. Push opening brackets
3. Match closing brackets with stack top
4. Validate final stack emptiness
```

### Time Complexity Breakdown:
- **Character Processing**: O(N) time where N=string length
- **Stack Operations**: O(1) per push/pop operation
- **Total**: O(N) time, O(N) space in worst case
- **Optimal**: Cannot do better than processing each character

### Alternative Approaches:

#### 1. Counter Replacement (O(N) time, O(1) space)
```go
func isValidCounters(s string) bool {
    // Only works for single bracket type
    // Not suitable for multiple bracket types
    // ... implementation details omitted
}
```
- **Pros**: Constant space
- **Cons**: Only works for single bracket type

#### 2. Two-Pass Validation (O(N) time, O(N) space)
```go
func isValidTwoPass(s string) bool {
    // First pass: check character counts
    // Second pass: validate order
    // More complex than stack approach
    // ... implementation details omitted
}
```
- **Pros**: Can provide more detailed error info
- **Cons**: More complex, same asymptotic complexity

#### 3. Recursive Approach (O(N) time, O(N) space)
```go
func isValidRecursive(s string) bool {
    // Recursively find and remove matched pairs
    // Less efficient than stack
    // ... implementation details omitted
}
```
- **Pros**: Elegant for some variations
- **Cons**: Less efficient, potential stack overflow

### Extensions for Interviews:
- **Different Bracket Types**: Handle angle brackets, quotes
- **Error Reporting**: Return specific error positions/types
- **Multiple Validation**: Validate multiple strings efficiently
- **Performance Analysis**: Discuss worst-case scenarios
- **Real-world Applications**: HTML/XML validation, expression parsing
*/
func main() {
	// Test cases
	testCases := []string{
		"()",
		"()[]{}",
		"(]",
		"([)]",
		"{[]}",
		"",
		"(",
		")",
		"({[]})",
		"({[)]})",
		"((()))",
		"{{{{}}}}",
		"({})",
		"[({})]",
		"([{}])",
		"([{}]))",
	}
	
	for i, s := range testCases {
		result := isValid(s)
		fmt.Printf("Test Case %d: \"%s\" -> Valid: %t\n", i+1, s, result)
	}
}

File: long_questions/DSA/Leetcode/java/25_String_Parsing_Stacks/020_Valid_Parentheses.java
import java.util.*;

public class ValidParentheses {
    
    // 20. Valid Parentheses
    // Time: O(N), Space: O(N)
    public static boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        
        // Map closing brackets to opening brackets
        Map<Character, Character> bracketMap = new HashMap<>();
        bracketMap.put(')', '(');
        bracketMap.put('}', '{');
        bracketMap.put(']', '[');
        
        for (char c : s.toCharArray()) {
            // If it's an opening bracket, push to stack
            switch (c) {
                case '(', '{', '[':
                    stack.push(c);
                    break;
                case ')', '}', ']':
                    // If stack is empty or top doesn't match, return false
                    if (stack.isEmpty() || stack.pop() != bracketMap.get(c)) {
                        return false;
                    }
                    break;
                default:
                    // Invalid character
                    return false;
            }
        }
        
        // Stack should be empty for valid string
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        // Test cases
        String[] testCases = {
            "()",
            "()[]{}",
            "(]",
            "([)]",
            "{[]}",
            "",
            "(",
            ")",
            "({[]})",
            "({[)]})",
            "((()))",
            "{{{{}}}}",
            "({})",
            "[({})]",
            "([{}])",
            "([{}]))"
        };
        
        for (int i = 0; i < testCases.length; i++) {
            boolean result = isValid(testCases[i]);
            System.out.printf("Test Case %d: \"%s\" -> Valid: %b\n", 
                i + 1, testCases[i], result);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: String Parsing with Stacks
- **Bracket Matching**: Validate balanced parentheses/brackets
- **Stack Operations**: Push opening, pop for closing
- **Character Mapping**: Map closing brackets to opening ones
- **LIFO Processing**: Last opened must be first closed

## 2. PROBLEM CHARACTERISTICS
- **String Validation**: Check if string has valid bracket sequence
- **Multiple Bracket Types**: Parentheses, braces, brackets
- **Balanced Property**: Every opening has matching closing
- **Order Constraint**: Closing must match most recent opening

## 3. SIMILAR PROBLEMS
- Valid Palindrome
- Remove Invalid Parentheses
- Longest Valid Parentheses
- Minimum Add to Make Valid

## 4. KEY OBSERVATIONS
- Stack naturally handles LIFO bracket matching
- Map closing brackets to opening for validation
- Stack empty at end means all brackets matched
- Any mismatched or extra closing makes string invalid
- Time complexity: O(N) where N is string length

## 5. VARIATIONS & EXTENSIONS
- Different bracket types
- Minimum insertions to balance
- Maximum nesting depth
- Bracket coloring problems
- Expression evaluation

## 6. INTERVIEW INSIGHTS
- Clarify: "What characters should I consider?"
- Edge cases: empty string, single bracket, mismatched types
- Time complexity: O(N) vs O(N²) naive
- Space complexity: O(N) worst case

## 7. COMMON MISTAKES
- Not handling all bracket types
- Incorrect mapping of closing to opening
- Not checking stack emptiness properly
- Forgetting to handle invalid characters
- Off-by-one errors in string processing

## 8. OPTIMIZATION STRATEGIES
- Early termination on obvious mismatches
- Use switch statement for character handling
- Handle invalid characters gracefully
- Optimize for specific character sets

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like matching socks and shoes:**
- You have a pile of socks (opening brackets)
- Each time you encounter a sock, put it on the pile
- When you encounter a shoe (closing bracket), it must match the sock
- Take the top sock from the pile and pair with the shoe
- If shoe doesn't match top sock, or pile is empty when shoe arrives → invalid
- At the end, pile should be empty (all socks paired)

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: String containing brackets
2. **Goal**: Check if brackets are balanced
3. **Output**: Boolean indicating validity

#### Phase 2: Key Insight Recognition
- **"What makes brackets balanced?"** → Each opening has matching closing
- **"How to track openings?"** → Stack (LIFO) data structure
- **"What's the matching rule?"** → Closing must match most recent opening
- **"Why stack?"** → Natural LIFO behavior for bracket matching

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use a stack:
1. Map closing brackets to opening ones
2. For each character in string:
   - If opening: push to stack
   - If closing: check stack top matches
   - If matches: pop from stack
   - If doesn't match: return false
3. At end, stack should be empty for valid string"
```

#### Phase 4: Edge Case Handling
- **Empty string**: Return true (no brackets to match)
- **Single bracket**: Return false (unmatched)
- **Mismatched types**: Return false immediately
- **Extra closing**: Return false when stack empty

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
String: "({[]})"

Human thinking:
"Let's process character by character:

Stack: [], Position: 0
Char '(': opening → push '(', Stack: ['(']
Char '{': opening → push '{', Stack: ['(', '{']
Char '[': opening → push '[', Stack: ['(', '{', '[']
Char ']': closing → top is '[', matches! → pop, Stack: ['(', '{']
Char ']': closing → top is '{', doesn't match! → return false ✗

Found mismatch: string is invalid ✓"
```

#### Phase 6: Intuition Validation
- **Why it works**: Stack ensures LIFO matching property
- **Why it's efficient**: Each character processed once
- **Why it's correct**: All bracket pairs validated

### Common Human Pitfalls & How to Avoid Them
1. **"Why not count brackets?"** → Can't handle order, only count
2. **"What about recursion?"** → Stack is more natural for this problem
3. **"How to handle multiple types?"** → Need proper mapping
4. **"What about invalid characters?"** → Should return false

### Real-World Analogy
**Like matching containers and lids:**
- You have containers (opening brackets) and lids (closing brackets)
- Each container must match its specific lid type
- You stack containers as you encounter them
- When you find a lid, it must match the top container
- If lid doesn't match container type, you can't seal it
- At the end, all containers should be sealed (stack empty)
- This ensures proper container-lid matching

### Human-Readable Pseudocode
```
function isValid(s):
    stack = empty stack
    bracketMap = {')': '(', '}': '{', ']': '['}
    
    for character in s:
        if character is opening bracket:
            stack.push(character)
        else if character is closing bracket:
            if stack is empty or stack.pop() != bracketMap[character]:
                return false
    
    return stack.isEmpty()
```

### Execution Visualization

### Example: "({[]})"
```
Character Processing:
Stack Evolution:
Position 0: '(', push → ['(']
Position 1: '{', push → ['(', '{']
Position 2: '[', push → ['(', '{', '[']
Position 3: ']', top='[' matches → ['(', '{']
Position 4: ']', top='{' matches → ['(']
Position 5: '}', top='(' matches → []
Position 6: ')', stack empty → VIOLATION!

Result: Invalid (unmatched ')') ✓

Visualization:
( { [ ] } )
↑ ↑ ↑ ↑ ↑
| | | | |
| | | | └─ Mismatch!
```

### Key Visualization Points:
- **Stack behavior** naturally handles LIFO matching
- **Bracket mapping** ensures correct pairing
- **Early termination** on mismatches
- **Final check** ensures all brackets matched

### Memory Layout Visualization:
```
Stack Evolution:
Step 1: push '(' → ['(']
Step 2: push '{' → ['(', '{']
Step 3: push '[' → ['(', '{', '[']
Step 4: pop '[' (matches ']') → ['(', '{']
Step 5: pop '{' (matches ']') → ['(']
Step 6: pop '(' (matches ')') → []
Step 7: encounter ')' with empty stack → ERROR

Final State: Empty stack would be valid
Actual: Stack empty when encountering ')' → Invalid
```

### Time Complexity Breakdown:
- **Each Character**: O(1) stack operations
- **String Processing**: O(N) where N is string length
- **Total**: O(N) time, O(N) worst-case space
- **Optimal**: Cannot do better than O(N) for this problem
- **vs Counting**: O(N) time but can't handle order
*/
